Feature: Mythic Keystone dungeon support

tdaugaard opened the pull request:
Re-based and fixed the old PR on top of the new main branch and redid the commits to make more logical sense, also removed some stuff that really wasn't specific to this feature and thus shouldn't go in this PR.

I have done extensive testing on this branch when doing M+ and there's only one thing I have yet to figure out how to fix, and that is that _all_ dungeons _except_ Karazhan are almost exactly on point with the VideoSegment timing, but _both_ Upper and Lower Karazhan are almost always off by at least 7-10 seconds. So basically, a boss segment starts, but you're 7-10 seconds into the boss fight. I think it's okay to leave it like that for now, as you can just press arrow left one time to go back 30 seconds and watch from there, until I can figure out exactly why it's happening for that dungeon.

This will fix issue #43

aza547 commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+    // add a new boss encounter video segment.
+    if (recorder.isRecording && activeChallengeMode) {
+        const vSegment = new ChallengeModeVideoSegment(
+            VideoSegmentType.BossEncounter, eventDate, getRelativeTimestampForVideoSegment(eventDate),
```
can we spread these args over some new lines for readability

aza547 commented on src/main/logutils.ts:
```diff
@@ -416,13 +519,25 @@ function handleEncounterStartLine (line: LogLine): void {
+        )
+
+        // Add a trash segment as the boss encounter ended
+        activeChallengeMode.addVideoSegment(vSegment, videoStopDate);
```
What about dungeons where bosses into boss is pulled. I'm almost certain this happens with some routes.

tdaugaard replied:
I've done a lot of dungeons and I've never seen a boss being pulled _into_ a boss. Do you have an example?

aza547 replied:
I had Taza'vesh in mind but I'm not sure as I basically never do M+.

 As per earlier happy enough if there is a few rough edges like this here.

tdaugaard replied:
Especially Taza'vesh will have a skip before the first boss then, and although it's not a complicated skip by any stretch of imagination, I'd personally like to be able to review it if we screw it up :p 

The actual reason for having a first timeline segment before the first boss is just to simply be able to seek to 0:00 withuot having to use the video player seek bar.

aza547 replied:
Ah so it would just show like "trash" 0-30s but there would be no combat in that, just running into position? Actually I like that as it is then. I agree that's important stuff you want to see and be able to skip to!

aza547 commented on src/main/logutils.ts:
```diff
@@ -575,6 +696,32 @@ function zoneChangeStop (line: LogLine): void {
     recorder.stop(metadata);
 }
 
+/**
```
Not sure how I feel about this. What about horribly bad runs with like 20 deaths, is that all going to appear in the context menu? Might get a bit busy. What about hunter feign deaths which causes a UNIT_DIED event to fire?

tdaugaard replied:
> What about horribly bad runs with like 20 deaths, is that all going to appear in the context menu?

We will be better and die less! Just kidding, I get what you mean - especially on raid bosses where several players might die in close succession either because of mechanics of due to a wipe being called.

I _could_ make some kind of logic such that a player death would only count if there has been no other player deaths in X number of seconds, or simply ignore player deaths after a certain number.

For now, player deaths aren't exposed in the UI - I created the logic for it but ultimately decided to leave it out due to space issues - we'll talk about what to do with it after M+ is merged in, as discussed on Discord too. 

Does that sound OK?

> What about hunter feign deaths which causes a UNIT_DIED event to fire?

Good point! I can actually see here, that the very last value in the `UNIT_DIED` line indicates if the unit is "unconscious" (i.e. feigned death) so I can/will filter on that. Shouldn't have feign deaths included.

aza547 replied:
Oh, that's awesome on the feign death thing. I had no idea that was available. That'd be great to include. 

Ultimately I think I don't care about this now I know it's not ending up in the context menu!

aza547 commented on src/renderer/Layout.tsx:
```diff
@@ -250,11 +255,18 @@ export default function Layout() {
     const video = state.videoState[category][state.videoIndex];
     const videoFullPath = video.fullPath;
     const key = "videoPanel" + index;
+    const isMythicPlus = video.category === VideoCategory.MythicPlus && video.challengeMode !== undefined
```
some brackets would make this more readable

aza547 commented on src/renderer/Layout.tsx:
```diff
@@ -250,11 +255,18 @@ export default function Layout() {
+    const isMythicPlus = video.category === VideoCategory.MythicPlus && video.challengeMode !== undefined
+    let videoPoster = readyPoster;
+
+    // Show a poster of the dungeon for M+ instead of "Select a video"
```
That's cool... and maybe is reason for having larger resolutions. Would like to do this for other categories eventually.

tdaugaard replied:
Yeah, the larger images take up quite a bit of space but I made them large so that they could be scaled down a bit if you wanted that. It looks bad to scale something _up_.

I was thinking about updating the whole `assets/dungeon/`  directory, but since many of the SL dungeons won't be in a rotation for a long time, it seemed moot.

aza547 commented on src/renderer/VideoButton.tsx:
```diff
@@ -42,26 +40,35 @@ export default function VideoButton(props: any) {
   // BGs don't log COMBATANT_INFO events so we can't display a lot of stuff
   // that we can for other categories. 
   const isBG = category === VideoCategory.Battlegrounds;
+  const isMythicPlus = category == VideoCategory.MythicPlus && video.challengeMode !== undefined;
```
brackets please!

aza547 commented on src/renderer/images.ts:
```diff
@@ -24,28 +26,31 @@ const arenaIDs = [
 
 let dungeon: imageObject = {};
 
+try {
```
empty try block?

tdaugaard replied:
Oversight - it shouldn't be there.

aza547 reviewed (changes requested):
- I thought video files should be 200x100 to be in-line with what we have already. I thought having large files causes the video button to display a corner of them. That being said it looks OK, so maybe you made some CSS changes that I've yet to see. 

- I don't think I'm a huge fan of the loading screen images (the WoW logo and text can overlap our text which I think looks bad. 

- The text is a different font/size to other buttons still. Can we make it consistent? 

- Logging is pretty ugly for the final metadata event, can we split so it's not a one line JSON string but a properly formatted JSON? 

I had to rush a little as getting dragged to a birthday dinner so my review of the renderer/ files was a bit limited in depth. 

Obviously this is a huge bit of work and will be awesome function to have, so thanks again for all your efforts. We should also re-advertise (as many of requested) when this is in (and confirmed to be stable, probably).

aza547 commented on src/renderer/VideoButton.tsx:
```diff
@@ -102,14 +109,105 @@ export default function VideoButton(props: any) {
+     */
+    const videoSegments = video.challengeMode.videoSegments.map((segment: any) => {
+      let videoSegmentMenu;
+      let segmentDurationText
```
semi colons on a few lines here and below

tdaugaard replied:
I have fixed it better and utilized the fact that `.push()` can take an arbitrary number of elements.

tdaugaard replied:
I have fixed it better and utilized the fact that `.push()` can take an arbitrary number of elements.

aza547 commented on src/renderer/rendererutils.ts:
```diff
@@ -36,22 +19,26 @@ const getInstanceDifficulty = (difficultyID: number): InstanceDifficultyType | n
 } 
 
+const getVideoResult = (video: any): boolean => {
+    if (video.challengeMode !== undefined) {
```
Why does this need special handling? It is still a boolean at the end of the day?

tdaugaard replied:
You mean why ChallengeMode need its own handling or why I'm testing for `undefined`?

aza547 replied:
Both I suppose. Why isn't video.result === video.challengeMode.timed in this context?

tdaugaard replied:
Because a dungeon can be completed, but not timed. If a dungeon isn't completed, `video.result` will be false, if it's completed by not timed `video.result = true` and `video.challengeMode.timed = false`.

aza547 replied:
Obvious really... thanks!

aza547 reviewed (commented):
What about timestamps in the test log (I commented on this in the other PR)?

tdaugaard commented on src/main/constants.ts:
```diff
@@ -142,14 +146,208 @@ const raids: { [id: number]: string; } = {
 
+/**
+ * Dungeons by map Id
+ * Names have been shortened, or abbreviated due to size constraints i
```
There was missing an 'n' from the end. 🤣

tdaugaard commented on src/renderer/VideoButton.tsx:
```diff
@@ -102,14 +109,105 @@ export default function VideoButton(props: any) {
+
+    // If the video metadata doesn't contain its own alloted time array, we'll use whichever
+    // is current in the constants.
+    const keystoneAllottedTime = video.challengeMode.allottedTime ?? dungeonTimersByMapId[video.challengeMode.mapId];
```
> Best guess is to handle the case if M+ timers change in a patch, we can still be accurate for previous videos?

Precisely. Timers are occassionally, but rarely, adjusted in the beginning when a new dungeon enters the rotation. I'm neutral as to include this or not.

tdaugaard commented on src/renderer/VideoButton.tsx:
```diff
@@ -102,14 +109,105 @@ export default function VideoButton(props: any) {
+      keystoneAllottedTime,
+      video.challengeMode.duration
+    );
+    keystonePlusses = '+'.repeat(keystoneResult)
```
`keystoneUpgradeLevel` - which is what it signifies - value 0 - 3.

tdaugaard commented:
> What about timestamps in the test log (I commented on this in the other PR)?

I don't think we should design the app around the tests.

I've a log replay script in the works which can "replay" a given log file, adjusting the timestamps on the fly such that this wouldn't be a problem.

aza547 commented:
> > What about timestamps in the test log (I commented on this in the other PR)?
> 
> I don't think we should design the app around the tests.
> 
> I've a log replay script in the works which can "replay" a given log file, adjusting the timestamps on the fly such that this wouldn't be a problem.

Maybe I wasn't clear - I'm just suggesting we tweak the timestamps in the dungeon.txt file which is part of the tests to make the seeking testable here. So a tweak to a test file, not a functional change.

tdaugaard commented:
> Maybe I wasn't clear - I'm just suggesting we tweak the timestamps in the dungeon.txt file which is part of the tests to make the seeking testable here. So a tweak to a test file, not a functional change.

I probably misunderstood - I will make a new test file with adjusted timestamps. :)

aza547 commented:
> > Maybe I wasn't clear - I'm just suggesting we tweak the timestamps in the dungeon.txt file which is part of the tests to make the seeking testable here. So a tweak to a test file, not a functional change.
> 
> I probably misunderstood - I will make a new test file with adjusted timestamps. :)

I had in mind some crappy interpolation to take number of lines in that test file, divide the 10 (seconds) by that and then re-create them all equally spaced throughout the 10s. I think that's pretty easy and good enough for test purposes (if you'd like to leave this to me, feel free).

tdaugaard commented:
> I thought video files should be 200x100 to be in-line with what we have already. I thought having large files causes the video button to display a corner of them. That being said it looks OK, so maybe you made some CSS changes that I've yet to see.

When using the loading screen images as the 'poster' for the video player, it wouldn't look nice to scale up a 200x100 picture. I've set `background-size: cover` for the `.videoButton.dungeon` selector.

> I don't think I'm a huge fan of the loading screen images (the WoW logo and text can overlap our text which I think looks bad.

Not sure what you mean with overlap - can you show me an example screenshot? Is it on the video button or as the video player 'poster' you don't like?

> The text is a different font/size to other buttons still. Can we make it consistent?

I've changed `.zone` text to be 16px (up from 13px) because the zone element is used for the key level (e.g. "++17"). `zone` class is used for both the dungeon name and the keystone upgrade & level.

I just want it to look good; I don't really care about whether it's 13px or 16px, but 13px seems to be too small - would it be OK to change it to 16px for everything?

> Logging is pretty ugly for the final metadata event, can we split so it's not a one line JSON string but a properly formatted JSON?

Agreed - done! 2 space indentation.

aza547 commented:
> > I thought video files should be 200x100 to be in-line with what we have already. I thought having large files causes the video button to display a corner of them. That being said it looks OK, so maybe you made some CSS changes that I've yet to see.
> 
> When using the loading screen images as the 'poster' for the video player, it wouldn't look nice to scale up a 200x100 picture. I've set `background-size: cover` for the `.videoButton.dungeon` selector.
> 
> > I don't think I'm a huge fan of the loading screen images (the WoW logo and text can overlap our text which I think looks bad.
> 
> Not sure what you mean with overlap - can you show me an example screenshot? Is it on the video button or as the video player 'poster' you don't like?

![Uploading image.png…]()

> 
> > The text is a different font/size to other buttons still. Can we make it consistent?
> 
> I've changed `.zone` text to be 16px (up from 13px) because the zone element is used for the key level (e.g. "++17"). `zone` class is used for both the dungeon name and the keystone upgrade & level.
> 
> I just want it to look good; I don't really care about whether it's 13px or 16px, but 13px seems to be too small - would it be OK to change it to 16px for everything?
> 
> > Logging is pretty ugly for the final metadata event, can we split so it's not a one line JSON string but a properly formatted JSON?
> 
> Agreed - done! 2 space indentation.

aza547 commented:
The text that I don't like is mostly the bits at the bottom FYI, the name and the timed/depleted indicator

tdaugaard commented:
> Agree, but I scaled down those pictures initially. No reason we can't get the high-res versions again for this in the future.

The pictures are large, and have to be large, because they are used as 'poster' image for the video player.
I personally like them and have no problem with the text overlapping, or rather overlaying, the background though.

Do you want me to:

1. Find non-loading screen looking images for the dungeons instead
2. Remove the 'poster' feature altogether and use non-loading screen looking low-res pictures
3. Keep the 'poster' feature and use different images for the video player 'poster' (current large ones) and the VideoButton, with the button using a low-res, non-loading screen looking, version of them

Let me know :)

tdaugaard commented:
> The text that I don't like is mostly the bits at the bottom FYI, the name and the timed/depleted indicator

The size? It can be larger. I did make the player name smaller for M+ as I consider that information unimportant for a dungeon run, but acknowledge that it might be more important in the PvP categories.

aza547 commented:
> > The text that I don't like is mostly the bits at the bottom FYI, the name and the timed/depleted indicator
> 
> The size? It can be larger. I did make the player name smaller for M+ as I consider that information unimportant for a dungeon run, but acknowledge that it might be more important in the PvP categories.

Yeah the size. I think for people playing multiple chars this info is super useful. I'd also prefer it to stay the same size across different categories - if you could revert that :)

tdaugaard commented on src/main/keystone.ts:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    /**
+     * Add a player death to the log
```
`playerDeaths` has been moved to  `Metadata` type and the comment has been rewritten.

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -39,6 +40,7 @@ const interestingCombatLogEvents: InterestingCombatLogEventsType = {
     'ZONE_CHANGE': handleZoneChange,
     'COMBATANT_INFO': handleCombatantInfoLine,
     'SPELL_AURA_APPLIED': handleSpellAuraAppliedLine,
+    'UNIT_DIED': handleUnitDiedLine,
```
Commented about this elsewhere. There's a flag in UNIT_DIED combat log line (last one, number 9) that indicates this.

aza547 commented:
> > Agree, but I scaled down those pictures initially. No reason we can't get the high-res versions again for this in the future.
> 
> The pictures are large, and have to be large, because they are used as 'poster' image for the video player. I personally like them and have no problem with the text overlapping, or rather overlaying, the background though.
> 
> Do you want me to:
> 
> 1. Find non-loading screen looking images for the dungeons instead
> 2. Remove the 'poster' feature altogether and use non-loading screen looking low-res pictures
> 3. Keep the 'poster' feature and use different images for the video player 'poster' (current large ones) and the VideoButton, with the button using a low-res, non-loading screen looking, version of them
> 
> Let me know :)

I think my preference is non-loading screen images, no problem with the big images - agree with the reasoning for that. This isn't really a big deal. I probably want to re-style these buttons some point soon to look better as I think it's one of the weaker bits aesthetically.

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+    const zoneName = line.args[2];
+    const mapId = parseInt(line.args[3], 10);
+
+    if (!(mapId in dungeonTimersByMapId && mapId in dungeonsByMapId)) {
```
> This is a bit ugly, can we split it out?

I'm generally against making variables that are only used in a single following comparison, but if that's what you want, I'll do it.

> Was also suprised to see no exit from this function here, don't you need to return to avoid processing a dungeon we don't recognise?

Yes and no. I was hoping to and planning on making it so that in the future when new dungeons are added, we wouldn't avoid recording and saving videos from unknown dungeons, encounters and such so that we could simply update the app and the details would show up after the recording had been saved.

`handleEncounterStartLine()` also doesn't care about the encounter existing or not.

aza547 replied:
Ah, I am of the mindset that if it makes code readable then it's a good thing. I appreciate that's an opinion not a fact so I'm happy either way.

tdaugaard replied:
I split it. It's your project afterall :)

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+    )
+
+    activeChallengeMode.addVideoSegment(new ChallengeModeVideoSegment(
+        VideoSegmentType.Trash, videoStartDate, 0
```
_Theater of Pain_ comes to mind as one such dungeon where it was a regular thing to skip the initial 3 mobs, and bring the 4th on top of the boss - besides that I don't recall any dungeons where doing two bosses without any trash is possible; HOWEVER even if that was the case there would probably be a skip strategy that would be relevant to have a beginning timestamp of.

I could label the first segment as 'Intro' segment instead, if you want?

aza547 replied:
If it's just a case of labelling then I don't think I care too much -- If it's functionally a non-issue and the bottom line is we just occasionally label some strats wrongly then that seems like an easy fix to make when people complain about it rather than worry now.

tdaugaard replied:
It's just labeling. I guess the "real fix" here is to make it so that if a timeline segment (video segment) is too short to have value, e.g. 10 seconds, then it'll be discarded and whatever comes next (like a boss) will take its place, but I think those refinements are better to keep for future improvements.

aza547 replied:
Awesome - totally happy with that.

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+
+    // The actual log duration of the dungeon, from which keystone upgrade
+    // levels can be calculated
+    activeChallengeMode.duration = Math.round(parseInt(line.args[4], 10) / 1000);
```
I don't know if it is, but the duration as taken here from argument 4 is the duration that the game considers for the keystone completion upgrade levels; I didn't want to take the risk that the (end - start) log line timestamps would be off by even 500 ms, as that _can_ mean the difference between timing a key or not timing it.

aza547 replied:
OK, can we just say this in a comment? Sounds fine, but good to have for future understanding

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+    const duration = Math.round(milliSeconds / 1000);
+
+    metadata.duration = duration + overrun;
+    metadata.result = Boolean(parseInt(line.args[1]));
```
I think I mentioned that elsewhere;

`metadata.result` = keystone completed?
`metadata.challengeMode.timed` = if completed, was it timed?

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+    // Calculate whether the key was timed or not
+    activeChallengeMode.timed = calculateKeystoneCompletionResult(activeChallengeMode.allottedTime, activeChallengeMode.duration) > 0;
+
+    console.debug("[ChallengeMode] Ending current video segment")
```
Adding additional logging - you are right, but it is doing both things - ending the video segment _and_ the CM instance.
```typescript
console.debug("[ChallengeMode] Starting Challenge Mode instance")
```

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+    console.debug("[ChallengeMode] Ending current video segment")
+    activeChallengeMode.endVideoSegment(videoStopDate);
+
+    // If last video segment is less than 10 seconds long, discard it.
```
Perhaps I could rename `VideoSegment` to `TimelineSegment`? I agree it could be confusing.

aza547 replied:
I think I prefer videoSegment, but not overly fussed - would be happy with a comment here

tdaugaard replied:
I renamed the whole video segment thing because I did find it confusing myself but did't have a better alternative at the time. It's now called timeline segments. 🤷‍♂️

tdaugaard commented on src/main/logutils.ts:
```diff
@@ -373,32 +376,132 @@ const determineArenaMatchResult = (line: LogLine): any[] => {
+        return 0;
+    }
+
+    return (currentDate.getTime() - videoStartDate.getTime()) / 1000;
```
Yes we could!

aza547 replied:
(to be clear - fine if you don't do this now - just a comment for future!)

tdaugaard commented:
> Yeah the size. I think for people playing multiple chars this info is super useful. I'd also prefer it to stay the same size across different categories - if you could revert that :)

Gonna be a very tight fit though. I could change "Timed" and "Depleted" to "T" and "D" or something else; I don't know.

![image](https://user-images.githubusercontent.com/875107/190912431-e590325b-579e-4785-8f42-3d0ff0e86a94.png)

Example of a maximum length character name (12 characters) - it doesn't look great at all:

![image](https://user-images.githubusercontent.com/875107/190912545-224280bc-79a2-4096-a920-c0df6d38bc2c.png)

aza547 commented:
> > Yeah the size. I think for people playing multiple chars this info is super useful. I'd also prefer it to stay the same size across different categories - if you could revert that :)
> 
> Gonna be a very tight fit though. I could change "Timed" and "Depleted" to "T" and "D" or something else; I don't know.
> 
> ![image](https://user-images.githubusercontent.com/875107/190912431-e590325b-579e-4785-8f42-3d0ff0e86a94.png)
> 
> Example of a maximum length character name (12 characters) - it doesn't look great at all:
> 
> ![image](https://user-images.githubusercontent.com/875107/190912545-224280bc-79a2-4096-a920-c0df6d38bc2c.png)

Time / Depl would be OK imo?

tdaugaard commented:
> Time / Depl would be OK imo?

Sounds good to me.

> Or even better +1, +2, +3 or "Depl"?

Actually, that's what the "+", "++", "+++" signifies in the keystone level - I thought it was a good way to show it as that is how some people write it when talking about expectations (in dungeon finder).

What do you prefer?

aza547 commented:
> > Time / Depl would be OK imo?
> 
> Sounds good to me.
> 
> > Or even better +1, +2, +3 or "Depl"?
> 
> Actually, that's what the "+", "++", "+++" signifies in the keystone level - I thought it was a good way to show it as that is how some people write it when talking about expectations (in dungeon finder).
> 
> What do you prefer?

Oh I see! I wasn't aware that was a convention (and maybe others wouldn't be either). I think I do prefer the +N style.

tdaugaard commented:
Honestly, I think this looks better too:

![image](https://user-images.githubusercontent.com/875107/190916408-d73ef7a4-3081-496a-a361-19306b42d220.png)

tdaugaard commented:
I renamed everything related to "video segments" to be "timeline segments" to avoid confusing it for physical video data.

tdaugaard commented on src/main/keystone.ts:
```diff
@@ -0,0 +1,127 @@
+        }
+    }
+
+    return 0;
```
> Could use an enum for this rather than assigning numbers.

An enum does not make sense here, in my opinion, since keystone upgrade levels is a numeric value by design and this method reflects that. I don't know how I should make it an enum even.

> I don't think this shows in the UI. Could we make the UI show +3, +2, +1 or depleted, if we already have the info?

Yes, that's been addressed in a newer commit regarding the look of the videobutton.

aza547 replied:
Yeah fair enough, happy with that.

aza547 reviewed (approved):
Thanks for all the hard work here. Excited to see this feature go live! 

I'll let you do the honors and hit merge :)
